src/Compilers/Core/Portable/SourceGeneration/RunResults.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Text;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis
+{
+    public readonly struct GeneratorDriverRunResult
+    {
+        internal GeneratorDriverRunResult(ImmutableArray<GeneratorRunResult> results, ImmutableArray<SyntaxTree> trees, ImmutableArray<Diagnostic> diagnostics)
+        {
+            this.Results = results;
+            this.SyntaxTrees = trees;
+            this.Diagnostics = diagnostics;
+        }
+
+        public ImmutableArray<GeneratorRunResult> Results { get; }
+
+        public ImmutableArray<Diagnostic> Diagnostics { get; }
+
+        public ImmutableArray<SyntaxTree> SyntaxTrees { get; }
+    }
+
+    public readonly struct GeneratorRunResult
+    {
+        internal GeneratorRunResult(ISourceGenerator generator, ImmutableArray<GeneratedSourceResult> generatedTrees, ImmutableArray<Diagnostic> diagnostics, Exception? exception)
+        {
+            this.Generator = generator;
+            this.GeneratedTrees = generatedTrees;
+            this.Diagnostics = diagnostics;
+            this.Exception = exception;
+        }
+
+        public ISourceGenerator Generator { get; }
+
+        public ImmutableArray<GeneratedSourceResult> GeneratedTrees { get; }
+
+        public ImmutableArray<Diagnostic> Diagnostics { get; }
+
+        public Exception? Exception { get; }
+
+        public bool Succeeded => Exception is null;
+    }
+
+    public readonly struct GeneratedSourceResult
```

Comment:
One future potential direction for this feature is generating non-source files along with source files. If that happens will we regret using `Source` in the name here? Or would the natural extension be to include a `GeneratedAdditionalFilesResult`?

Reply:
Yeah, I think we'd have a another set of `Generated<Whatever>Result` on the parent struct, rather than trying to represent multiple things here.

---
In reply to: [460214330](https://github.com/dotnet/roslyn/pull/46296#discussion_r460214330) [](ancestors = 460214330)